iroh-net/src/dns/node_info.rs

```diff
@@ -0,0 +1,267 @@
+//! This module contains functions and structs to lookup node information from DNS
+//! and to encode node information in Pkarr signed packets.
+
+use std::{collections::HashMap, fmt, str::FromStr};
+
+use anyhow::{anyhow, bail, Result};
+use hickory_proto::error::ProtoError;
+use hickory_resolver::{Name, TokioAsyncResolver};
+use url::Url;
+
+use crate::{key::SecretKey, AddrInfo, NodeAddr, NodeId};
+
+const ATTR_RELAY: &str = "relay";
+const ATTR_NODE_ID: &str = "node";
+
+/// The label for the node info TXT record
+pub const IROH_NODE_TXT_LABEL: &str = "_iroh_node";
+
+/// Lookup node info by domain name
+///
+/// The domain name must either contain an _iroh_node TXT record or be a CNAME record that leads to
+/// an _iroh_node TXT record.
+pub async fn lookup_by_domain(resolver: &TokioAsyncResolver, domain: &str) -> Result<NodeAddr> {
+    let name = Name::from_str(domain)?;
+    let info = lookup_node_info(resolver, name).await?;
+    Ok(info.into())
+}
+
+/// Lookup node info by node id and origin domain name.
+pub async fn lookup_by_id(
+    resolver: &TokioAsyncResolver,
+    node_id: &NodeId,
+    origin: &str,
+) -> Result<NodeAddr> {
+    let domain = format!("{}.{}", to_z32(node_id), origin);
+    lookup_by_domain(resolver, &domain).await
+}
+
+pub(crate) async fn lookup_node_info(
+    resolver: &TokioAsyncResolver,
+    name: Name,
+) -> Result<NodeInfo> {
+    let name = ensure_iroh_node_txt_label(name)?;
+    let lookup = resolver.txt_lookup(name).await?;
+    NodeInfo::from_hickory_lookup(lookup.as_lookup())
+}
+
+fn ensure_iroh_node_txt_label(name: Name) -> Result<Name, ProtoError> {
+    if name.iter().next() == Some(IROH_NODE_TXT_LABEL.as_bytes()) {
+        Ok(name)
+    } else {
+        Name::parse(IROH_NODE_TXT_LABEL, Some(&name))
+    }
+}
+
+/// Encode a [`NodeId`] in [`z-base-32`] encoding.
+///
+/// [z-base-32]: https://philzimmermann.com/docs/human-oriented-base-32-encoding.txt
+pub fn to_z32(node_id: &NodeId) -> String {
+    z32::encode(node_id.as_bytes())
```

Comment:
At the risk of opening a giant rabbit hole or the mother of all bike shed discussions: Can we just make the string representation of our node ids be zbase32 as well?

I think that this is going to become a source of confusion especially when people start manually debugging the DNS records using nslookup/dig. And there is really nothing wrong with zbase32. It's not like our encoding (lower case base32) is completely standard either.

Also there are some valid reasons why the z encoding is better in the admittedly rare cases where people will type those by hand. https://philzimmermann.com/docs/human-oriented-base-32-encoding.txt

I am not going to argue about using hex again, since the 63 char subdomain fragment limit is a very convincing argument...

Reply:
I agree that it would be much nicer to have the same encoding everywhere, and that a different encoding for the DNS names than how we print NodeIds in general is quite unfortunate.

* The cleanest would be switch all our string encodings from lowerbase32 to zbase32. This would mean, though, that with the next release all existing tickets, hashes, nodeids would break. How much pain would that cause? I'm not sure. Maybe yes? Depends on how and where and in which ways Iroh is already in production use I guess?
* If we switched only the `to_string()/from_string()` of `NodeId`, we wouldn't have much breakage I think, however then we would move the inconsistency from "two ways to encode a NodeId" towards "NodeIds and hashes are encoded with different formats", which is also not great.

Reply:
lets not do any of this as part of this PR, please make a new issue if you want to bikeshed encodings, and we can consider this independently